File: benchmarks/SearchEngine.CompetitorBenchmarks/minisearch/glob-match.mjs
export function globMatch(pattern, word) {
  const p = [...pattern];
  const w = [...word];
  let patternIndex = 0;
  let wordIndex = 0;
  let starPatternIndex = -1;
  let starWordIndex = 0;

  while (wordIndex < w.length) {
    if (patternIndex < p.length) {
      const patternChar = p[patternIndex];
      if (patternChar === "*") {
        starPatternIndex = patternIndex;
        starWordIndex = wordIndex;
        patternIndex++;
        while (patternIndex < p.length && p[patternIndex] === "*") patternIndex++;
        continue;
      }
      if (patternChar === "?" || patternChar === w[wordIndex]) {
        patternIndex++;
        wordIndex++;
        continue;
      }
    }
    if (starPatternIndex >= 0) {
      starWordIndex++;
      wordIndex = starWordIndex;
      patternIndex = starPatternIndex + 1;
      while (patternIndex < p.length && p[patternIndex] === "*") patternIndex++;
      continue;
    }
    return false;
  }

  while (patternIndex < p.length && p[patternIndex] === "*") patternIndex++;
  return patternIndex === p.length;
}

File: benchmarks/SearchEngine.CompetitorBenchmarks/tantivy/src/glob_match.rs
//! Glob matcher aligned with SearchEngine.Sharp `GlobMatcher` (anchored full-string match).

pub fn glob_match(pattern: &str, word: &str) -> bool {
    let pattern = pattern.as_bytes();
    let word = word.as_bytes();
    let mut pattern_index = 0usize;
    let mut word_index = 0usize;
    let mut star_pattern_index: isize = -1;
    let mut star_word_index = 0usize;

    while word_index < word.len() {
        if pattern_index < pattern.len() {
            let pattern_char = pattern[pattern_index];
            if pattern_char == b'*' {
                star_pattern_index = pattern_index as isize;
                star_word_index = word_index;
                pattern_index += 1;
                while pattern_index < pattern.len() && pattern[pattern_index] == b'*' {
                    pattern_index += 1;
                }
                continue;
            }
            if pattern_char == b'?' || pattern_char == word[word_index] {
                pattern_index += 1;
                word_index += 1;
                continue;
            }
        }

        if star_pattern_index >= 0 {
            star_word_index += 1;
            word_index = star_word_index;
            pattern_index = (star_pattern_index + 1) as usize;
            while pattern_index < pattern.len() && pattern[pattern_index] == b'*' {
                pattern_index += 1;
            }
            continue;
        }

        return false;
    }

    while pattern_index < pattern.len() && pattern[pattern_index] == b'*' {
        pattern_index += 1;
    }

    pattern_index == pattern.len()
}
